pkg/provision/primitives/public_ip.go

```diff
@@ -26,11 +34,122 @@ func (p *Provisioner) publicIPProvision(ctx context.Context, reservation *provis
 }
 
 func (p *Provisioner) publicIPProvisionImpl(ctx context.Context, reservation *provision.Reservation) (result PublicIPResult, err error) {
-	return PublicIPResult{}, nil
+	config := PublicIP{}
+
+	network := stubs.NewNetworkerStub(p.zbus)
+
+	if err := json.Unmarshal(reservation.Data, &config); err != nil {
+		return PublicIPResult{}, errors.Wrap(err, "failed to decode reservation schema")
+	}
+
+	pubIP6Base, err := network.GetPublicIPv6Subnet()
+	if err != nil {
+		return PublicIPResult{}, errors.Wrap(err, "could not look up ipv6 prefix")
+	}
+
+	tapName := fmt.Sprintf("p-%s", reservation.ID) // TODO: clean this up, needs to come form networkd
+	fName := filterName(reservation.ID)
+	mac := ifaceutil.HardwareAddrFromInputBytes(config.IP.IP.To4())
+
+	predictedIPv6, err := predictedSlaac(pubIP6Base.IP, mac.String())
+	if err != nil {
+		return PublicIPResult{}, errors.Wrap(err, "could not look up ipv6 prefix")
+	}
+
+	err = setupFilters(ctx, fName, tapName, config.IP.IP.To4().String(), predictedIPv6, mac.String())
+	return PublicIPResult{
+		ID: reservation.ID,
+		IP: config.IP.IP.String(),
+	}, err
 }
 
 func (p *Provisioner) publicIPDecomission(ctx context.Context, reservation *provision.Reservation) error {
 	// Disconnect the public interface from the network if one exists
 	network := stubs.NewNetworkerStub(p.zbus)
+	fName := filterName(reservation.ID)
+	if err := teardownFilters(ctx, fName); err != nil {
+		log.Error().Err(err).Msg("could not remove filter rules")
+	}
 	return network.DisconnectPubTap(reservation.ID)
 }
+
+func filterName(reservationID string) string {
+	return fmt.Sprintf("r-%s", reservationID)
+}
+
+func setupFilters(ctx context.Context, fName string, iface string, ip string, ipv6 string, mac string) error {
+	cmd := exec.CommandContext(ctx, "/bin/sh", "-c",
```

Comment:
Would it be nicer if we use go `text/template` to generate a template and write it directly to `nft` stdin, then just execute `nft -f -`

Reply:
Not sure. Keep in mind that we are not just setting a template, but actually applying rules to an existing nft setup. I personally don't know nft enough to know if this is possible